core/src/main/kotlin/org/jetbrains/kotlinx/dataframe/impl/io/readJson.kt

```diff
@@ -189,9 +201,21 @@ internal fun fromJsonListAnyColumns(
                         )
                     }
 
-                    "NaN" -> {
-                        nanIndices.add(i)
-                        collector.add(null)
+                    is JsonPrimitive -> {
```

Comment:
While this `xOrNull` conversion works, it's quite heavy as it works by throwing/catching exceptions. Ilya actually recently replaced similar logic from the notebook kernel because it was so heavy. It was replaced by Jackson. Now I know that's not an option here, but maybe we could try another approach if that's possible.

Reply:
I might be missing something
Could you please explain in more detail or provide an example from the notebook?

Reply:
Sure, let's say we want to parse an array of `null`s. In this case, for each element it would check [`intOrNull`](https://github.com/Kotlin/kotlinx.serialization/blob/d2dc7d2978bd62971e8ab00de197d59d7e378ed2/formats/json/commonMain/src/kotlinx/serialization/json/JsonElement.kt#L267), `longOrNull`, `doubleOrNull`, and `floatOrNull`. All of these work with `mapExceptionsToNull {}`, meaning, for each element in the array an exception is thrown and caught 4 times. This is very heavy, especially when it happens very often.

Reply:
Hm, indeed this isn't the good place that should be fixed when supporting Serializable for DataFrame and DataRow. I also think that part of this might be optimized at runtime since our value is already wrapped in JsonPrimitive. There are doubts that jdk8 handles such optimizations well

Reply:
I moved the null check higher up. Given the current capabilities of kotlinx-serialization, there are two options: either leave it as it is now, or write a custom serializer. However, the logic in the custom serializer would be quite similar. We can always take a String and then manually attempt to convert it to the desired type. This might be slightly better in terms of performance, and we would need to account for all edge cases for different types